src/tiledbsoma_ml/query_ids.py

```diff
@@ -0,0 +1,198 @@
+
+        rng = np.random.default_rng(seed)
+        shuffled_joinids = rng.permutation(obs_joinids)
+
+        if method == "deterministic":
+            split_idxs = np.round(split_fracs * n_obs).astype(int)
+        elif method == "multinomial":
+            split_idxs = np.cumsum(rng.multinomial(n_obs, np.array(fracs)))
+        elif method == "stochastic_rounding":
+            fracs_arr = np.array(fracs) * n_obs
+            split_bases = fracs_arr.astype(int)
+            split_idxs = split_bases.cumsum()
+            remainder_fracs = fracs_arr - split_bases
+            n = len(fracs)
+            remainders = np.zeros(n, dtype=int)
+            total_remainders = round(np.sum(remainder_fracs))
+            while total_remainders > 0:
+                pvals = remainder_fracs / np.sum(remainder_fracs)
+                choice = rng.choice(n, p=pvals)
+                remainders[choice] += 1
+                remainder_fracs[choice] = 0
+                total_remainders -= 1
+            split_idxs += remainders.cumsum()
+        else:
+            raise ValueError(f"Unknown sampling method: {method}")
+
+        splits = [
+            np.sort(split)
+            for split in np.array_split(shuffled_joinids, split_idxs[:-1])
+        ]
+        return tuple(evolve(self, obs_joinids=split) for split in splits)
+
+    def partitioned(
+        self,
+        partition: Optional[Partition] = None,
+    ) -> "QueryIDs":
+        """Create a new |QueryIDs| with |Q.obs_joinids| corresponding to a given GPU/worker |Partition|.
+
+        If ``None`` is provided, world size, rank, num workers, and worker ID will be inferred using helper functions
+        that read env vars (see |get_distributed_world_rank|, |get_worker_world_rank|).
+
+        When ``WORLD_SIZE > 1``, each GPU will receive the same number of samples (meaning up to ``WORLD_SIZE-1``
+        samples may be dropped).
+        """
+        if self.partition:
+            raise ValueError(
+                f"QueryIDs has already been partitioned ({self.partition})"
+            )
+
+        obs_joinids = self.obs_joinids
+        if partition:
+            rank = partition.rank
+            world_size = partition.world_size
+            worker_id = partition.worker_id
+            n_workers = partition.n_workers
+        else:
+            rank, world_size = get_distributed_world_rank()
+            worker_id, n_workers = get_worker_world_rank()
+            partition = Partition(
+                rank=rank,
+                world_size=world_size,
+                worker_id=worker_id,
+                n_workers=n_workers,
+            )
+
+        gpu_splits = splits(len(obs_joinids), world_size)
+        gpu_split = obs_joinids[gpu_splits[rank] : gpu_splits[rank + 1]]
+
+        # Trim all GPU splits to be of equal length (equivalent to a "drop_last"); required for distributed training.
+        # TODO: may need to add an option to do padding as well.
```

Comment:
it would be good to add the "pad" option to backlog, as this is very commonly used as well (it accomplishes the same thing by padding out the partial/final mini-batches, rather than throwing away the partial bits)

Reply:
There are examples of how this should be done in the PyTorch ecosystem.  IIRC, the typical approach is to just duplicate the contents of the contents of the too-small minibatch until it is large enough.  I.e., add duplicates of existing data.

E.g., if your mini batch size is 8, and you end up with [0,1,2,3,4,5], then just duplicate the last two, yielding [0,1,2,3,4,5,4,5]

Reply:
Also, if you elect not to do this in this PR, which is fine, it would be good to capture this as a story. I'm pretty sure the request will come eventually.

Reply:
Just confirming this was captured as a story/issue so we don't lose track of it.

Reply:
Filed as [sc-63968](https://app.shortcut.com/tiledb-inc/story/63968/tiledbsoma-ml-support-padding-final-mini-batch)